### PR TITLE
Fix Spearbit gas optimization issues

### PR DESCRIPTION
**Pull request by mikeyrf**
Addresses gas optimization issues:

- [x] [Retrieve roles via constants in import (13)](https://github.com/spearbit-audits/overlay/issues/13)
- [x] [Put risk parameters in an array (14)](https://github.com/spearbit-audits/overlay/issues/14)
- [x] [Replace square with multiplication (15)](https://github.com/spearbit-audits/overlay/issues/15)
- [x] [Redundant Math.min() (20)](https://github.com/spearbit-audits/overlay/issues/20)
- [x] [Optimize power functions (21)](https://github.com/spearbit-audits/overlay/issues/21)
- [x] [Remove unused / redundant functions and variables (22)](https://github.com/spearbit-audits/overlay/issues/22)
- [x] [Optimizations if data.macroWindow is constant (23)](https://github.com/spearbit-audits/overlay/issues/23)
- [x] [Simplify circuitBreaker (24)](https://github.com/spearbit-audits/overlay/issues/24)
- [x] [No need to cache immutable values (27)](https://github.com/spearbit-audits/overlay/issues/27)
- [x] [Replace memory with calldata (28)](https://github.com/spearbit-audits/overlay/issues/28)
- [x] [Store result of expression in temporary variable (32)](https://github.com/spearbit-audits/overlay/issues/32)
- [x] [Set pos.entryPrice to 0 after liquidation (33)](https://github.com/spearbit-audits/overlay/issues/33)
- [x] [Simplify _midFromFeed() (35)](https://github.com/spearbit-audits/overlay/issues/35)
- [x] [Use implicit truncation of timestamp (38)](https://github.com/spearbit-audits/overlay/issues/38)

### CODE REVIEW

File: contracts/OverlayV1Factory.sol [hunk 1]
```diff
@@ -140,6 +138,13 @@ contract OverlayV1Factory is IOverlayV1Factory {
         require(IOverlayV1FeedFactory(feedFactory).isFeed(feed), "OVLV1: feed does not exist");
     }
 
+    /// @notice Checks all risk params are within acceptable bounds
+    function _checkRiskParams(uint256[14] calldata params) private {
+        for (uint256 i = 0; i < params.length; i++) {
```
Comment (hrkrshnn) [inline contracts/OverlayV1Factory.sol:143]: I'd cache the length:
```solidity
uint length = params.length;
for (uint i = 0; i < length; i++) { ...}
```
would save around `3` gas per iteration.

File: contracts/OverlayV1Factory.sol [hunk 2]
```diff
@@ -170,259 +140,38 @@ contract OverlayV1Factory is IOverlayV1Factory {
...
-                params.priceDriftUpperLimit <= MAX_PRICE_DRIFT_UPPER_LIMIT,
-            "OVLV1: priceDriftUpperLimit out of bounds"
+            value >= minValue && value <= maxValue,
+            string(
+                abi.encodePacked(
+                    "OVLV1: param ",
+                    Strings.toString(uint256(name)),
+                    " out of bounds"
+                )
+            )
```
[the remark this answers is not in the document]
Comment (mikeyrf) [reply contracts/OverlayV1Factory.sol:160]: Ah ok! Wasn't aware of `string.concat()`. Will replace.

File: contracts/OverlayV1Factory.sol [hunk 3]
```diff
@@ -11,82 +13,48 @@ import "./libraries/Risk.sol";
...
-        uint256 tradingFeeRate
-    );
-    event MinimumCollateralUpdated(
-        address indexed user,
-        address indexed market,
-        uint256 minCollateral
-    );
-    event PriceDriftUpperLimitUpdated(
+    // risk param bounds
+    uint256[14] public PARAMS_MIN = [
```
[the remark this answers is not in the document]
Comment (mikeyrf) [reply contracts/OverlayV1Factory.sol:19]: Yea the risk parameters ideally would only be set at deployment, then never changed again. In practice, hoping for once every few months re-evaluating risk parameters.

I do like how concise the code has become with the enum and these arrays. Feels much easier to read.

From checking gas estimates with deploying the new version of the `OverlayV1Factory`, it's not too terrible as far as I can tell. About 6.36M to deploy the factory and ~40k each time governance needs to set a risk parameter.

```
OverlayV1Factory <Contract>
   ├─ constructor     -  avg: 6358549  avg (confirmed): 6358549  low: 6358543  high: 6358555
   ├─ deployMarket    -  avg: 2085560  avg (confirmed): 3933479  low:   27570  high: 3947803
   ├─ addFeedFactory  -  avg:   40970  avg (confirmed):   47457  low:   24294  high:   47457
   ├─ setRiskParam    -  avg:   34821  avg (confirmed):   39181  low:   23199  high:   54968
   └─ setFeeRecipient -  avg:   26681  avg (confirmed):   31564  low:   24144  high:   31564
OverlayV1FeedFactoryMock <Contract>
   ├─ constructor     -  avg:  455532  avg (confirmed):  455532  low:  455532  high:  455532
   └─ deployFeed      -  avg:  272856  avg (confirmed):  292094  low:   22802  high:  292094
OverlayV1Token <Contract>
   ├─ constructor     -  avg: 1540957  avg (confirmed): 1540957  low: 1540957  high: 1540957
   ├─ grantRole       -  avg:  103374  avg (confirmed):  103374  low:   95683  high:  111067
   ├─ mint            -  avg:   66749  avg (confirmed):   66749  low:   66749  high:   66749
   └─ transfer        -  avg:   43586  avg (confirmed):   43586  low:   36082  high:   51094
OverlayV1UniswapV3Factory <Contract>
   ├─ constructor     -  avg: 2404287  avg (confirmed): 2404287  low: 2404287  high: 2404287
   └─ deployFeed      -  avg: 1108951  avg (confirmed): 1651350  low:   24153  high: 1651350
```

File: contracts/OverlayV1Market.sol [hunk 4]
```diff
@@ -121,6 +125,7 @@ contract OverlayV1Market is IOverlayV1Market {
 
         // set the risk params
         for (uint256 i = 0; i < _params.length; i++) {
+            _cacheRiskCalc(Risk.Parameters(i), _params[i]);
```
[the remark this answers is not in the document]
Comment (mikeyrf) [reply contracts/OverlayV1Market.sol:128]: Fair point. Will rewrite.

File: contracts/OverlayV1Factory.sol (hunk 1 above)
Comment (mikeyrf) [reply contracts/OverlayV1Factory.sol:143]: Sounds good!

**Pull request merged**